### DISCUSSION TITLE
Nvidia Reflex implementation

### DISCUSSION THREAD

**Discussion by GGamer6458** [ppy/osu#23274] [Ideas] [answered]
Have you thought about implementing Nvidia Reflex within the game?

Nvidia Reflex helps reduce latency by reducing the rendering queue as much as possible, by syncing the CPU and the GPU with each other.

If you want more information on this subject, there's a video from Nvidia themselves explaining Nvidia Reflex:
https://www.youtube.com/watch?v=-cXg7GQogAE

**Comment by peppy** [comment, marked as answer]
This is planned.

**Reply by luvm0cha** [reply]
NVIDIA Reflex 2 was announced 7 months ago. I dont see that as an option today to enable or disable but it would be a nice feature to add.

**Reply by Joehuu** [reply]
Draft PR in https://github.com/ppy/osu/pull/35678.

**Comment by GhostylGit** [comment]
Something similar to NVIDIA Reflex 2 could be implemented by disabling the rendering of the cursor, and then once a frame has been rendered, adding in the cursor on top of the frame as a post-processing effect using the latest input information from the input thread instead.

Also, something separate that you could do to reduce stuttering/animation error is during frametime spikes you could display the previous frame again but with an up to date cursor position so you never _feel_ a stutter _with your cursor_ (i.e. you would always have 1 frame displayed every 1ms no matter what). And or, you could use it for frame pacing by delaying displaying frames, by say 1ms, and then equally spacing them out so they are displayed at an extremely consistent rate, with 0ms of animation error, but then have the lowest latency _feeling_ input possible with up to date input, but this would have the downside of having 1ms higher _visual_ latency of everything except your cursor.

Obviously none of this effects _real_ latency like Reflex 1 does but I speculate that it would feel better to play with, especially just the first part I suggested here combined with Reflex 1. [Here's more info on NVIDIA's suggested Reflex 2](https://www.nvidia.com/en-us/geforce/news/reflex-2-even-lower-latency-gameplay-with-frame-warp/) but an implementation into osu! would be different due to it being a 2d game and being multithreaded, but this is the same concept.
